include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -5706,14 +5719,168 @@ FEEvaluationAccess<dim, dim, Number, is_face, VectorizedArrayType>::operator=(
+
+          for (unsigned int comp = 0; comp < n_components; ++comp)
+            value_out[comp] = this->values_quad[comp * nqp + q_point] *
+                              jac[comp][comp] *
+                              inv_det; // / this->jacobian[0][comp][comp];
+        }
+
+      // Affine or general cell
+      else
+        {
+          const Tensor<2, dim, dealii::VectorizedArray<Number>> &inv_t_jac =
+            (this->cell_type > internal::MatrixFreeFunctions::affine) ?
+              this->jacobian[q_point] :
+              this->jacobian[0];
+          const Tensor<2, dim, VectorizedArrayType> &t_jac = invert(inv_t_jac);
+
+          // Derivatives are reordered for faces. Need to take this into account
+          const std::uint8_t        fn = (is_face) ? this->get_face_no() : 0;
+          const VectorizedArrayType inv_det = (is_face && dim == 2 && fn < 2) ?
+                                                -determinant(inv_t_jac) :
+                                                determinant(inv_t_jac);
+          const unsigned int *index = internal::get_index<dim, is_face>(fn);
+          // J * u * det(J^-1)
+          for (unsigned int comp = 0; comp < n_components; ++comp)
+            {
+              value_out[comp] =
+                this->values_quad[q_point] * t_jac[index[0]][comp] * inv_det;
+              for (unsigned int e = 1; e < dim; ++e)
+                value_out[comp] += this->values_quad[e * nqp + q_point] *
+                                   t_jac[index[e]][comp] * inv_det;
+            }
+        }
+      return value_out;
+    }
+  else
+    {
+      return BaseClass::get_value(q_point);
+    }
+}
 
 template <int dim, typename Number, bool is_face, typename VectorizedArrayType>
 inline DEAL_II_ALWAYS_INLINE Tensor<2, dim, VectorizedArrayType>
 FEEvaluationAccess<dim, dim, Number, is_face, VectorizedArrayType>::
   get_gradient(const unsigned int q_point) const
 {
-  return BaseClass::get_gradient(q_point);
+  // Check if Piola transform is required
+  if (this->data->element_type ==
+      internal::MatrixFreeFunctions::ElementType::tensor_raviart_thomas)
+    {
+#  ifdef DEBUG
+      Assert(this->gradients_quad_initialized == true,
+             internal::ExcAccessToUninitializedField());
+#  endif
+
+      AssertIndexRange(q_point, this->n_quadrature_points);
+      Assert(this->jacobian != nullptr,
+             internal::ExcMatrixFreeAccessToUninitializedMappingField(
+               "update_gradients"));
+      const std::size_t nqp = this->n_quadrature_points;
+      Tensor<1, dim, Tensor<1, dim, VectorizedArrayType>> grad_out;
+
+      // Cartesian cell
+      if (!is_face &&
+          this->cell_type == internal::MatrixFreeFunctions::cartesian)
+        {
+          const Tensor<2, dim, VectorizedArrayType> &inv_t_jac =
+            this->jacobian[0];
+          const Tensor<2, dim, VectorizedArrayType> &t_jac = invert(inv_t_jac);
```

Comment:
This is not documented anywhere, but you do not need to compute inverse matrix for Cartesian/affine *cells*
```suggestion
          const Tensor<2, dim, VectorizedArrayType> &t_jac = this->jacobian[1];
```
This is filled in here for cells: https://github.com/dealii/dealii/blob/7e1892d3381e93ad2af8d5c2f3a51af23bef2ac1/include/deal.II/matrix_free/mapping_info.templates.h#L1212-L1217. Note that this is not done for faces (if I remember correctly) but we could do it if necessary).

Reply:
Ok! It does seem unnecessary to compute it twice... I'll have a look at it tomorrow (and also update the documentation)

Reply:
I can't find that this is used anywhere in `matrix_free/fe_evalluation.h` but I'm unsure if it is used anywhere else. Because I would like to change it to the transpose jacobian instead of just the jacobian. That way there's no need to call `transpose(invert(inv_t_jac))` if the cell is general in for instance `get_values`. I would also add it for faces.

Reply:
The use is here: https://github.com/dealii/dealii/blob/c02086aecc5e6588047865e2bb4ad058bf95e346/include/deal.II/matrix_free/fe_evaluation_data.h#L1276-L1292
Regarding a call to `transpose`, I would expect that an optimizing compiler optimizes the memory operations away. (It does not always work, but I've seen it work often.) Hence, it should not matter whether you call `transpose(jac)`. Note that for general cells the Jacobians are not stored, just for the affine/Cartesian case (as the latter would be too memory intensive).